driver-core/src/main/com/mongodb/internal/connection/DnsMultiServerCluster.java

```diff
@@ -58,13 +60,27 @@ public void initialize(final Collection<ServerAddress> hosts) {
             }
 
             private Collection<ServerAddress> applySrvMaxHosts(final Collection<ServerAddress> hosts) {
-                Collection<ServerAddress> newHosts = hosts;
+                Collection<ServerAddress> newHosts = new ArrayList<>(hosts);
+                List<ServerAddress> existingHostList = DnsMultiServerCluster.this.getCurrentDescription().getServerDescriptions()
+                        .stream().map(ServerDescription::getAddress).collect(Collectors.toList());
+                Integer numCurrentEndPoints = existingHostList.size();
+                Integer numRemovedEndPoints = 0;
+                Iterator<ServerAddress> iterator = existingHostList.iterator();
+                while (iterator.hasNext()) {
+                    ServerAddress host = iterator.next();
+                    if (!hosts.contains(host)) {
+                        iterator.remove();
+                        numRemovedEndPoints++;
+                    }
+                }
                 Integer srvMaxHosts = getSettings().getSrvMaxHosts();
                 if (srvMaxHosts != null && srvMaxHosts > 0) {
                     if (srvMaxHosts < hosts.size()) {
                         List<ServerAddress> newHostsList = new ArrayList<>(hosts);
```

Comment:
This is a bit confusing, because we have a newHosts above, but they have different meanings (one means, the new hosts based on the srv response, while the other means, hosts added in the srv response relative to prior hosts). We could call this `addedHosts` or `novelHosts` or something of the sort?